### ISSUE TITLE
How do I get the abundance of each query in `fastgather` and `multifastgather` results?

### ISSUE THREAD

**Issue by lhsnam** [sourmash-bio/sourmash_plugin_branchwater#163]
Hi,

I attempted to use 'sourmash_plugin_branchwater', which has a faster time benchmark than 'sourmash gather'. However, the 'gather' result only shows statistics for 'intersect_bp' (file listed below). Is this a trade-off for the shorter duration, or did I get something wrong?

Thank you,

[zr00210_1027.sig.gather.csv](https://github.com/sourmash-bio/sourmash_plugin_branchwater/files/13657467/zr00210_1027.sig.gather.csv)

**Comment by ctb** [comment]
yes! I tend to use the fastgather output as a picklist to `sourmash gather`, which gives the full output. Here are some command line examples: [link](https://github.com/sourmash-bio/sourmash_plugin_branchwater/blob/main/doc/README.md#using-fastgather-to-create-a-picklist-for-sourmash-gather)

It turns out we refer to this obliquely in the documentation, but nowhere do we state it clearly 😭 . Here's our current best shot:
>This code is still in prototype mode, and does not have all of the features of sourmash. As we add features we will move it back into the core sourmash code base; eventually, much of the code in this repository will be integrated into sourmash directly.

Please leave this issue open and I will update the text to be clearer about the situation. Thank you!

**Comment by lhsnam** [comment]
Oh! Thank you @ctb 

I noticed that in the Sourmash [manual page](https://sourmash.readthedocs.io/en/latest/sourmash-internals.html#much-faster-search-and-gather-with-branchwater) and [repo's issue #2827](https://github.com/sourmash-bio/sourmash/issues/2827), you mentioned this repository. I then applied with a few Fastq files and found that `fastmultigather` was much quicker at exporting results than both `gather` and `multigather`. I appreciate this repo!

---

In addition, I am looking for a way to shorten the duration of taxa profiling (many paired-end Fastq by a Genbank bacteria database >30Gb). That's why I approach this repo. May I have one more question clarifying how to work with sourmash (and this repo, too)?

So, from **sourmash**, `multigather` is now pre-loading the database and maps consecutively with the fastq files. Is it a better choice for me than `gather` each Fastq one by one? I don't really think that you mention is fully clear:
>"We don’t particularly suggest using `multigather`; we would prefer to make search databases faster."

Why pre-load the database and search consecutively with each Fastq slower?
While waiting for this **branchwater** plugin, what is the best solution for me to shorten the duration of taxa profiling, as I mentioned above?

Best,
Nam

**Comment by ctb** [comment]
> In addition, I am looking for a way to shorten the duration of taxa profiling (many paired-end Fastq by a Genbank bacteria database >30Gb). That's why I approach this repo. May I have one more question clarifying how to work with sourmash (and this repo, too)?

Sure! And thank you for asking ;)
 
> So, from **sourmash**, `multigather` is now pre-loading the database and maps consecutively with the fastq files. Is it a better choice for me than `gather` each Fastq one by one? I don't really think that you mention is fully clear:
> 
> > "We don’t particularly suggest using `multigather`; we would prefer to make search databases faster."

I'm clearly going to have to update the multigather documentation more...  my apologies!

Some key context is this sentence above [in that section of the docs](https://sourmash.readthedocs.io/en/latest/sourmash-internals.html#running-gather-many-times-multigather):
>In situations where loading the search database is slow...

It's hard to fully explain, since there's a lot of backstory and options and ..., but basically there are situations in which loading the database each time is a big cost, compared to doing the search. That is the case for doing small metagenome gathers, for example, or using certain database types.

But the crucial thing here is that for any real metagenome against any real database, the search itself is going to take far longer than the database load step, and so multigather doesn't offer you any real benefit compared to running gather. Moreover, gather is just much more convenient than multigather in terms of inputs and outputs and so on.

I am not sure I'm explaining that well but basically gather is just a much easier command to use in practice!

And, with fastgather, it's become so fast that this is even more true...
 
> Why pre-load the database and search consecutively with each Fastq slower? While waiting for this **branchwater** plugin, what is the best solution for me to shorten the duration of taxa profiling, as I mentioned above?

What I have taken to doing is exactly what you asked about above:
* use branchwater plugin/fastgather to generate results to use as a picklist
* then using sourmash gather with that picklist to generate "full" gather results

It's extremely fast and convenient!

Please ask more questions, and I'm also hoping to have some more information and scripts for you in the next week...

**Comment by lhsnam** [comment]
Sure, the more detailed information would be more helpful for users. It is clear to me now. That makes sense for my research task, too.

> * use branchwater plugin/fastgather to generate results to use as a picklist
> * then using sourmash gather with that picklist to generate "full" gather results

After applying your advice with the small database (gtdb-rs214-reps), I got a significantly quicker result and output similar to the regular `gather` I have often used. Now, I'm trying to run some samples with a bigger one - the GenBank database (36.8 GB). I hope that will work for me. I'll tell you about the benchmark later.

Best,
Nam

**Comment by ctb** [comment]
hi @lhsnam I've put together a functioning snakemake workflow that demonstrates the two steps involved - see rules `metag_fastgather` and `metag_gather` in the sourmash-slainte `Snakefile`:

https://github.com/dib-lab/sourmash-slainte/blob/0268dd5ba20c2449c363cbbd8fc7f8533105d41c/Snakefile#L239

Happy to explain what's going on in more detail if and as you're interested!

**Comment by lhsnam** [comment]
Hello @ctb, I hope you're doing well. I just wanted to follow up on our previous conversation regarding the benchmark. As I mentioned earlier, when I applied it to my Nextflow pipeline, I was able to achieve ~7 times faster results than using the original sourmash in any databases. I am thrilled with this improvement and feel that it will be good enough for me while waiting for the full development of this function. Thanks a lot for your advice!

Regarding the **sourmash-slainte**, it sounds fascinating to me. I will take some time to look into it and reach out to you in case of any confusion.
